src/trans.rs

```diff
@@ -226,9 +265,42 @@ impl<'v, 'tcx: 'v> BinaryenFnCtxt<'v, 'tcx> {
                     let expr = unsafe { BinaryenReturn(self.module, expr) };
                     binaryen_stmts.push(expr);
                 }
+                TerminatorKind::Switch { ref discr, .. } => {
+                    let adt = self.trans_lval(discr);
+                    let adt_ty = self.mir.lvalue_ty(*self.tcx, discr).to_ty(*self.tcx);
+
+                    if adt.offset.is_some() {
+                        panic!("unimplemented Switch with offset");
+                    }
+
+                    let adt_layout = self.type_layout(adt_ty);
+                    let discr_val = match *adt_layout {
+                        Layout::General { discr, .. } => {
+                            let discr_size = discr.size().bytes() as u32;
+                            debug!("emitting GetLocal({}) + Load for ADT Switch condition", adt.index.0);
+                            unsafe {
+                                let ptr = BinaryenGetLocal(self.module, adt.index, BinaryenInt32());
+                                BinaryenLoad(self.module, discr_size, 0, 0, 0, BinaryenInt32(), ptr)
+                            }
+                        }
+                        Layout::CEnum { .. } => {
+                            debug!("emitting GetLocal({}) for CEnum Switch condition", adt.index.0);
+                            unsafe {
+                                BinaryenGetLocal(self.module, adt.index, BinaryenInt32())
+                            }
+                        }
+                        _ => panic!("unimplemented discrimant value for Layout {:?}", adt_layout)
+                    };
+
+                    block_kind = BinaryenBlockKind::Switch(discr_val);
+                }
                 TerminatorKind::Call {
                     ref func, ref args, ref destination, ..
                 } => unsafe {
+                    // NOTE: plan for the calling convention: i32/i64 f32/f64 are to be passed using the wasm stack and function parameters.
+                    //       For the other types, the manual stack in linear memory will be used, and pointers into this stack
+                    //       passed as i32s. A call to a function returning a struct will require preparing the output return value space on
+                    //       the caller function's frame, and the called function will write its return value there to avoid memcpys
```

Comment:
Next time I'll do a quick pass over the code before I do the line-by-line pass so my questions make more sense :)

I don't think there's anything inherently wrong with just writing to the frame pointer, although that's still a little undefined. I'm assuming you mean something like, before the function call, the stack looks like:

```
... other stuff ...
-----
caller data (locals, etc)
-----
```

Then during the call it looks like
```
... other stuff ...
-----
caller data (locals, etc)
-----
callee data (locals, etc)
-----
```

And finally afterwards you have: 
```
... other stuff ...
-----
caller data (locals, etc)
-----
callee return value
-----
```

This works, but it can have some downsides. The callee has to move all of its own data out of the way before it returns (although really it can just reserve space for its return value before putting anything else on the stack). The caller will also probably have to move the value to the place where it wants it to be.

Alternatively, the caller can just pass in an output pointer saying "stick your return value here." That way when the function returns the data is already in the right place.

Hopefully that all makes sense. I opened #30 that we can use to discuss the design of this area some more.

Also, don't worry about being inexperienced. You're doing some fine compiler engineering work here!